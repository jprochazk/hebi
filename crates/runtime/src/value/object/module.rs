use indexmap::IndexMap;

use super::handle::Handle;
use super::{Access, Dict, Func, Str};

#[derive(Clone, Debug)]
pub struct Registry {
  pub modules: IndexMap<String, Module>,
}

impl Registry {
  pub fn new() -> Self {
    Self {
      modules: IndexMap::new(),
    }
  }
}

impl Default for Registry {
  fn default() -> Self {
    Self::new()
  }
}

impl Access for Registry {}

#[derive(Clone, Debug)]
pub struct Module {
  name: Str,
  main: Handle<Func>,
  pub globals: Handle<Dict>,
}

// TODO: get globals from module
impl Module {
  pub fn new(name: impl Into<Str>, main: Handle<Func>) -> Self {
    Self {
      name: name.into(),
      main,
      globals: Dict::new().into(),
    }
  }

  pub fn name(&self) -> &str {
    self.name.as_str()
  }

  pub fn main(&self) -> &Handle<Func> {
    &self.main
  }
}

impl Access for Module {}

#[derive(Clone, Debug)]
pub struct Path {
  segments: Vec<Str>,
}

impl Path {
  pub fn new(segments: Vec<Str>) -> Self {
    Self { segments }
  }

  pub fn segments(&self) -> &[Str] {
    &self.segments
  }
}

impl Access for Path {}
