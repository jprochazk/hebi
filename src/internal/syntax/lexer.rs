#![allow(non_camel_case_types)]

use std::borrow::Borrow;
use std::fmt;
use std::mem::{discriminant, take};
use std::ops::Range;

use logos::Logos;

use crate::span::Span;

#[derive(Clone, Debug)]
pub struct Token {
  pub ws: Option<u64>,
  pub kind: TokenKind,
  pub span: Span,
}

impl Token {
  pub fn is(&self, kind: impl Borrow<TokenKind>) -> bool {
    discriminant(&self.kind) == discriminant(kind.borrow())
  }
}

#[derive(Clone)]
pub struct Lexer<'src> {
  src: &'src str,
  inner: logos::Lexer<'src, TokenKind>,
  previous: Token,
  current: Token,
  ws: Option<u64>,
  eof: Token,
}

impl<'src> Lexer<'src> {
  pub fn new(src: &'src str) -> Self {
    let end = src.len();
    let eof = Token {
      ws: None,
      span: (end..end).into(),
      kind: TokenKind::Tok_Eof,
    };

    let mut lex = Self {
      src,
      inner: TokenKind::lexer(src),
      previous: eof.clone(),
      current: eof.clone(),
      ws: Some(0),
      eof,
    };
    lex.bump();

    lex
  }

  #[inline]
  pub fn previous(&self) -> &Token {
    &self.previous
  }

  #[inline]
  pub fn current(&self) -> &Token {
    &self.current
  }

  #[inline]
  pub fn lexeme(&self, token: &Token) -> &'src str {
    &self.src[Range::from(token.span)]
  }

  #[inline]
  pub fn bump(&mut self) {
    std::mem::swap(&mut self.previous, &mut self.current);

    self.current = self.next_token().unwrap_or(self.eof.clone());
  }

  fn next_token(&mut self) -> Option<Token> {
    let lexer = &mut self.inner;
    while let Some(kind) = lexer.next() {
      let lexeme = lexer.slice();
      let span = lexer.span().into();

      match kind {
        // Filter
        Ok(TokenKind::_Tok_Whitespace | TokenKind::_Tok_Comment) => continue,
        // Measure indentation
        Ok(TokenKind::_Tok_Indent) => {
          self.ws = Some(measure_indent(lexeme));
          continue;
        }
        // Return any other token
        Ok(kind) => {
          let token = Token {
            ws: take(&mut self.ws),
            kind,
            span,
          };
          return Some(token);
        }
        Err(_) => {
          let token = Token {
            ws: take(&mut self.ws),
            kind: TokenKind::Tok_Error,
            span,
          };
          return Some(token);
        }
      }
    }

    None
  }
}

// When adding a token, if it is matched using `token` directive only,
// then it should also be added to the `known` module below.
#[derive(Clone, Copy, Debug, Logos, PartialEq)]
pub enum TokenKind {
  // Keywords
  #[token("import")]
  Kw_Import,
  #[token("from")]
  Kw_From,
  #[token("as")]
  Kw_As,
  #[token("pub")]
  Kw_Pub,
  #[token("fn")]
  Kw_Fn,
  #[token("yield")]
  Kw_Yield,
  #[token("class")]
  Kw_Class,
  #[token("self")]
  Kw_Self,
  #[token("super")]
  Kw_Super,
  #[token("for")]
  Kw_For,
  #[token("is")]
  Kw_Is,
  #[token("in")]
  Kw_In,
  #[token("while")]
  Kw_While,
  #[token("loop")]
  Kw_Loop,
  #[token("return")]
  Kw_Return,
  #[token("break")]
  Kw_Break,
  #[token("continue")]
  Kw_Continue,
  #[token("print")]
  Kw_Print,
  #[token("if")]
  Kw_If,
  #[token("elif")]
  Kw_Elif,
  #[token("else")]
  Kw_Else,
  #[token("pass")]
  Kw_Pass,

  // Brackets
  #[token("{")]
  Brk_CurlyL,
  #[token("}")]
  Brk_CurlyR,
  #[token("(")]
  Brk_ParenL,
  #[token(")")]
  Brk_ParenR,
  #[token("[")]
  Brk_SquareL,
  #[token("]")]
  Brk_SquareR,

  // Misc characters
  #[token(".")]
  Op_Dot,
  #[token(",")]
  Tok_Comma,
  #[token(";")]
  Tok_Semicolon,
  #[token(":")]
  Tok_Colon,
  #[token("?")]
  Tok_Question,

  // Equals operators
  #[token("=")]
  Op_Equal,
  #[token("==")]
  Op_EqualEqual,
  #[token("+=")]
  Op_PlusEqual,
  #[token("-=")]
  Op_MinusEqual,
  #[token("/=")]
  Op_SlashEqual,
  #[token("*=")]
  Op_StarEqual,
  #[token("%=")]
  Op_PercentEqual,
  #[token("**=")]
  Op_StarStarEqual,
  #[token("??=")]
  Op_QuestionQuestionEqual,
  #[token("!=")]
  Op_BangEqual,
  #[token(":=")]
  Op_ColonEqual,

  // Operators
  #[token("+")]
  Op_Plus,
  #[token("-")]
  Op_Minus,
  #[token("/")]
  Op_Slash,
  #[token("*")]
  Op_Star,
  #[token("%")]
  Op_Percent,
  #[token("**")]
  Op_StarStar,
  #[token("??")]
  Op_QuestionQuestion,
  #[token("!")]
  Op_Bang,
  #[token(">")]
  Op_More,
  #[token(">=")]
  Op_MoreEqual,
  #[token("<")]
  Op_Less,
  #[token("<=")]
  Op_LessEqual,
  #[token("||")]
  Op_PipePipe,
  #[token("&&")]
  Op_AndAnd,
  #[token("..")]
  Op_Range,
  #[token("..=")]
  Op_RangeInc,

  // Literals
  /// `none`
  #[token("none")]
  Lit_None,
  #[regex("[0-9]([0-9_]*[0-9])?", priority = 10)]
  Lit_Int,
  /// `0`, `1.0`, `5e10`, etc.
  #[regex(r"[0-9]+(\.[0-9]+)?([Ee][+-]?[0-9]+)?")]
  Lit_Float,
  /// `true` or `false`
  #[token("true")]
  #[token("false")]
  Lit_Bool,
  #[regex(r#""([^"\\]|\\.)*""#)]
  Lit_String,
  /// `a`, `b_c`, `__x0`, etc.
  #[regex("[a-zA-Z_][a-zA-Z0-9_]*")]
  Lit_Ident,

  #[doc(hidden)]
  #[regex(r"(\r?\n)+[ ]*", priority = 10)]
  _Tok_Indent,
  #[doc(hidden)]
  #[regex(r"[ \n\r]+")]
  _Tok_Whitespace,
  #[doc(hidden)]
  #[regex(r"#[^\n]*")]
  _Tok_Comment,

  Tok_Error,
  Tok_Eof,
}

impl TokenKind {
  pub fn name(&self) -> &'static str {
    match self {
      TokenKind::Kw_Import => "import",
      TokenKind::Kw_From => "from",
      TokenKind::Kw_As => "as",
      TokenKind::Kw_Pub => "pub",
      TokenKind::Kw_Fn => "fn",
      TokenKind::Kw_Yield => "yield",
      TokenKind::Kw_Class => "class",
      TokenKind::Kw_Self => "self",
      TokenKind::Kw_Super => "super",
      TokenKind::Kw_For => "for",
      TokenKind::Kw_Is => "is",
      TokenKind::Kw_In => "in",
      TokenKind::Kw_While => "while",
      TokenKind::Kw_Loop => "loop",
      TokenKind::Kw_Return => "return",
      TokenKind::Kw_Break => "break",
      TokenKind::Kw_Continue => "continue",
      TokenKind::Kw_Print => "print",
      TokenKind::Kw_If => "if",
      TokenKind::Kw_Elif => "elif",
      TokenKind::Kw_Else => "else",
      TokenKind::Kw_Pass => "pass",
      TokenKind::Brk_CurlyL => "{",
      TokenKind::Brk_CurlyR => "}",
      TokenKind::Brk_ParenL => "(",
      TokenKind::Brk_ParenR => ")",
      TokenKind::Brk_SquareL => "[",
      TokenKind::Brk_SquareR => "]",
      TokenKind::Op_Dot => ".",
      TokenKind::Tok_Comma => ",",
      TokenKind::Tok_Semicolon => ";",
      TokenKind::Tok_Colon => ":",
      TokenKind::Tok_Question => "?",
      TokenKind::Op_Equal => "=",
      TokenKind::Op_EqualEqual => "==",
      TokenKind::Op_PlusEqual => "+=",
      TokenKind::Op_MinusEqual => "-=",
      TokenKind::Op_SlashEqual => "/=",
      TokenKind::Op_StarEqual => "*=",
      TokenKind::Op_PercentEqual => "%=",
      TokenKind::Op_StarStarEqual => "**=",
      TokenKind::Op_QuestionQuestionEqual => "?=",
      TokenKind::Op_BangEqual => "!=",
      TokenKind::Op_ColonEqual => ":=",
      TokenKind::Op_Plus => "+",
      TokenKind::Op_Minus => "-",
      TokenKind::Op_Slash => "/",
      TokenKind::Op_Star => "*",
      TokenKind::Op_Percent => "%",
      TokenKind::Op_StarStar => "**",
      TokenKind::Op_QuestionQuestion => "??",
      TokenKind::Op_Bang => "!",
      TokenKind::Op_More => ">",
      TokenKind::Op_MoreEqual => ">=",
      TokenKind::Op_Less => "<",
      TokenKind::Op_LessEqual => "<=",
      TokenKind::Op_PipePipe => "||",
      TokenKind::Op_AndAnd => "&&",
      TokenKind::Op_Range => "..",
      TokenKind::Op_RangeInc => "..=",
      TokenKind::Lit_None => "none",
      TokenKind::Lit_Int => "int",
      TokenKind::Lit_Float => "float",
      TokenKind::Lit_Bool => "bool",
      TokenKind::Lit_String => "string",
      TokenKind::Lit_Ident => "identifier",
      TokenKind::_Tok_Indent => "<indentation>",
      TokenKind::_Tok_Whitespace => "<whitespace>",
      TokenKind::_Tok_Comment => "<comment>",
      TokenKind::Tok_Error => "<error>",
      TokenKind::Tok_Eof => "<eof>",
    }
  }
}

fn measure_indent(s: &str) -> u64 {
  let pos = s.rfind('\n').unwrap_or(0);
  (s.len() - pos - 1) as u64
}

pub struct Tokens<'src>(pub Lexer<'src>);

impl<'src> Iterator for Tokens<'src> {
  type Item = (&'src str, Token);

  fn next(&mut self) -> Option<Self::Item> {
    let token = self.0.current().clone();
    self.0.bump();
    if !token.is(TokenKind::Tok_Eof) {
      Some((self.0.lexeme(&token), token))
    } else {
      None
    }
  }
}

pub struct DebugToken<'src>(pub Token, pub &'src str);
impl<'src> fmt::Debug for DebugToken<'src> {
  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
    let ws = self
      .0
      .ws
      .map(|v| v.to_string())
      .unwrap_or_else(|| "_".to_owned());
    let kind = self.0.kind;
    let span = self.0.span;
    let lexeme = self.1;
    if let TokenKind::Lit_Ident = self.0.kind {
      write!(f, "(>{ws} {kind:?} `{lexeme}` @{span})")
    } else {
      write!(f, "(>{ws} {kind:?} @{span})")
    }
  }
}

#[cfg(all(test, not(feature = "__miri")))]
mod tests;
