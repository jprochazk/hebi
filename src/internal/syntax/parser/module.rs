use super::*;

impl<'src> Parser<'src> {
  pub(super) fn module(mut self) -> Result<ast::Module<'src>, Vec<SpannedError>> {
    while !self.current().is(Tok_Eof) {
      if let Err(e) = self.top_level_stmt() {
        self.errors.push(e);
        self.sync();
      }
    }

    if !self.errors.is_empty() {
      return Err(self.errors);
    }

    Ok(self.module)
  }
}
