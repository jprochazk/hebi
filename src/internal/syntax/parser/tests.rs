use indoc::indoc;

use super::*;
use crate::internal::syntax::lexer::Lexer;
use crate::internal::vm::global::Global;

// TODO: emit input expression in snapshot
// do this for all snapshots tests that don't do it already

macro_rules! check_module {
  ($input:literal) => {
    let global = Global::default();
    let input = indoc!($input);
    match parse(global, input) {
      Ok(module) => assert_debug_snapshot!(module),
      Err(e) => {
        for err in e.errors() {
          eprintln!("{}", err.report(input, true));
        }
        panic!("Failed to parse source, see errors above.")
      }
    };
  };
  ($name:ident, $input:literal) => {
    #[test]
    fn $name() {
      check_module!($input);
    }
  };
}

macro_rules! check_expr {
  ($input:literal) => {
    let global = Global::default();
    let input = $input;
    match Parser::new(global, Lexer::new(input)).expr() {
      Ok(module) => assert_debug_snapshot!(module),
      Err(err) => {
        eprintln!("{}", err.report(input, true));
        panic!("Failed to parse source, see errors above.")
      }
    };
  };
}

macro_rules! check_error {
  ($input:literal) => {
    let global = Global::default();
    let input = indoc!($input);
    match parse(global, input) {
      Ok(_) => panic!("module parsed successfully"),
      Err(e) => {
        let mut errors = String::new();
        for err in e.errors() {
          errors += &err.report(input, true);
          errors += "\n";
        }
        assert_snapshot!(errors);
      }
    };
  };
  ($name:ident, $input:literal) => {
    #[test]
    fn $name() {
      check_error!($input);
    }
  };
}

#[test]
fn import_stmt() {
  check_module! {
    r#"#!hebi
      import module
      from module import z
      from module import x, y, z

      import module.nested
      from module.nested import z
      from module.nested import x, y, z
    "#
  };

  check_module! {
    r#"#!hebi
      import module as temp
      from module import z as temp
      from module import x as temp, y as temp, z as temp
      
      import module.nested as temp
      from module.nested import z as temp
      from module.nested import x as temp, y as temp, z as temp
    "#
  };

  check_error! {
    r#"#!hebi
      import a
        import b
    "#
  };

  check_error! {
    r#"#!hebi
      from m import a
        from m import b
    "#
  };

  check_error! {
    r#"#!hebi
      from
        m
    "#
  };

  check_error! {
    r#"#!hebi
      from m import
        a
    "#
  };

  check_error! {
    r#"#!hebi
      from m import a,
        b
    "#
  };

  check_error! {
    r#"#!hebi
      from m
        .b
    "#
  };
}

#[test]
fn binary_expr() {
  check_expr!(r#"a + b"#);
  check_expr!(r#"a - b"#);
  check_expr!(r#"a / b"#);
  check_expr!(r#"a ** b"#);
  check_expr!(r#"a * b"#);
  check_expr!(r#"a % b"#);
  check_expr!(r#"a == b"#);
  check_expr!(r#"a != b"#);
  check_expr!(r#"a > b"#);
  check_expr!(r#"a >= b"#);
  check_expr!(r#"a < b"#);
  check_expr!(r#"a <= b"#);
  check_expr!(r#"a is b"#);
  check_expr!(r#"a in b"#);
  check_expr!(r#"a && b"#);
  check_expr!(r#"a || b"#);
  check_expr!(r#"a ?? b"#);

  check_module! {
    r#"#!hebi
      a + b
      c + d
    "#
  };

  check_error! {
    r#"#!hebi
      a +
        b
    "#
  }
}

#[test]
fn unary_expr() {
  check_expr!(r#"+a"#);
  check_expr!(r#"-a"#);
  check_expr!(r#"!a"#);
  check_expr!(r#"?a.b[c].d()"#);
}

#[test]
fn postfix_expr() {
  check_expr!(r#"a.b[c].d"#);
  check_module! {
    r#"#!hebi
      a.b[c].d
      a.b[c].d
    "#
  };

  check_error! {
    r#"#!hebi
      a
      .b[c].d
    "#
  }
  check_error! {
    r#"#!hebi
      a.b[c]
      .d
    "#
  }
  check_error! {
    r#"a."#
  }
}

#[test]
fn call_expr() {
  check_expr!(r#"a(b, c,)"#);
  check_module! {
    r#"#!hebi
      a(b, c,)
      a(
        b,
      c, d
          ,
          )
    "#
  };
}

#[test]
fn simple_literal_expr() {
  check_module! {
    r#"#!hebi
      none
      true
      false
      1
      0.1
      1.5e3
      3.14e-3
      "\tas\\df\u{2800}\x28\n"
    "#
  }
}

#[test]
fn array_literal_expr() {
  check_module! {
    r#"#!hebi
      [0, 1, 2]
      [0,
       1,
       2,]
      [
        0,
        1,
        2,
      ]
    "#
  }
}

#[test]
fn table_literal_expr() {
  check_module! {
    r#"#!hebi
      {a:b, c:d, e:f}
      {a:b,
        c:d,
        e:f,}
      {
        a:b,
        c:d,
        e:f,
      }
    "#
  }

  check_module! {
    r#"#!hebi
      {[a]:b, [c]:d, [e]:f}
      {[a]:b,
       [c]:d,
       [e]:f,}
      {
        [a]:b,
        [c]:d,
        [e]:f,
      }
    "#
  }
}

#[test]
fn grouping_expr() {
  check_module! {
    r#"#!hebi
      # asdf
      (a + b)
      (
      a
       +
          b
      )
      (a
        (b))
      ((((a))))
    "#
  }
}

#[test]
fn assign_expr() {
  check_module! {
    r#"#!hebi
      # asdf
      a = b
      a := b
      a += b
      a -= b
      a /= b
      a *= b
      a %= b
      a **= b
      a ??= b
    "#
  }

  check_module! {
    r#"#!hebi
      # asdf
      a.b = b
      a.b += b
      a.b -= b
      a.b /= b
      a.b *= b
      a.b %= b
      a.b **= b
      a.b ??= b
    "#
  }

  check_error! {
    r#"#!hebi
      a
        = b
    "#
  }
  check_error! {
    r#"#!hebi
      a =
        b
    "#
  }

  check_error! {
    r#"a.b := c"#
  }

  check_error! {
    r#"a() = b"#
  }
}

#[test]
fn if_stmt() {
  check_module! {
    r#"#!hebi
      if a: pass
      elif b: pass
      elif c: pass
      else: pass
    "#
  }

  check_module! {
    r#"#!hebi
      if a:
        pass
      elif b:
        pass
      elif c:
        pass
      else:
        pass
    "#
  }

  check_module! {
    r#"#!hebi
      if a:
        a
        b
      elif b:
        a
        b
      elif c:
        a
        b
      else:
        a
        b
    "#
  }

  check_module! {
    r#"#!hebi
      if a:
        if b:
          pass
    "#
  }

  check_error! {
    r#"#!hebi
      if a:
        a
          b
      else: pass
    "#
  }

  check_error! {
    r#"#!hebi
      if a
        : pass
      else: pass
    "#
  }

  check_error! {
    r#"#!hebi
      if a: pass
      elif b
        : pass
      else: pass
    "#
  }

  check_error! {
    r#"#!hebi
      if a: pass
      elif b: pass
      else
        : pass
    "#
  }

  check_error! {
    r#"#!hebi
      if a: pass
      elif b: pass
        else: pass
    "#
  }

  check_error! {
    r#"#!hebi
      if a: pass
        elif b: pass
      else: pass
    "#
  }

  check_module! {
    r#"#!hebi
      if (
        some_very_long_condition &&
        with_many_sub_expressions() ||
        (which_are_complex.too())
      ):
        pass
    "#
  }
}

#[test]
fn loop_stmts() {
  check_module! {
    r#"#!hebi
      loop: pass
      loop:
        pass
    "#
  }

  check_error! {
    r#"#!hebi
      loop:
      pass
    "#
  }

  check_module! {
    r#"#!hebi
      while true: pass
      while true:
        pass
    "#
  }

  check_error! {
    r#"#!hebi
      while true:
      pass
    "#
  }

  check_module! {
    r#"#!hebi
      for i in iter(): pass
      for i in iter():
        pass
      for i in 0..10: pass
      for i in 0..10:
        pass
      for i in a()..b(): pass
      for i in a()..b():
        pass
      for i in 0..=10: pass
      for i in 0..=10:
        pass
      for i in a()..=b(): pass
      for i in a()..=b():
        pass
    "#
  }

  check_error! {
    r#"#!hebi
      for i in iter():
      pass
    "#
  }

  check_module! {
    r#"#!hebi
      loop:
        loop:
          a
          a
    "#
  }
}

#[test]
fn func_stmt() {
  check_module! {
    r#"#!hebi
      fn f(a, b, c,): pass
      fn f(a, b, c=d): pass
      fn f(a, b=c, d=e,): pass
    "#
  }

  check_error!(r#"fn f(a, b=c, d,): pass"#);
  check_error!(r#"fn f(*,): pass"#);
  check_error!(r#"fn f(**,): pass"#);
  check_error!(r#"fn f(**kwargs, a,): pass"#);
  check_error!(r#"fn f(a, b=,): pass"#);
  check_error!(r#"fn f(a, a): pass"#);
  check_error!(r#"fn f(a, *a): pass"#);
  check_error!(r#"fn f(a, *, a): pass"#);
  check_error!(r#"fn f(a, **a): pass"#);
  check_error! {
    r#"#!hebi
      fn f():
      pass
    "#
  }
  check_error! {
    r#"#!hebi
      fn():
        pass
    "#
  }
}

#[test]
fn ctrl_stmt() {
  check_module! {
    r#"#!hebi
      # simple
      loop:
        break
        continue
      
      fn f():
        yield
        yield v
        return
        return v

      # nested
      loop:
        loop:
          break
          continue
        break
        continue

      fn g():
        fn h():
          yield
          yield v
          return
          return v
        yield
        yield v
        return
        return v
      
      loop:
        fn i():
          yield
          yield v
          return
          return v
        break
        continue

      fn j():
        loop:
          break
          continue
        yield
        yield v
        return
        return v
      
      loop:
        fn k():
          loop:
            break
            continue
          yield
          yield v
          return
          return v
        break
        continue

      fn l():
        loop:
          fn m():
            yield
            yield v
            return
            return v
          break
          continue
        yield
        yield v
        return
        return v
    "#
  }

  check_error! {
    r#"#!hebi
      return v
    "#
  }

  check_error! {
    r#"#!hebi
      yield v
    "#
  }

  check_error! {
    r#"#!hebi
      continue
    "#
  }

  check_error! {
    r#"#!hebi
      break
    "#
  }
}

#[test]
fn print_stmt() {
  check_module! {
    r#"#!hebi
      print "a", 0, true
      print "a", 0, true
    "#
  }

  check_error! {
    r#"#!hebi
      print
        "a"
    "#
  }
}

#[test]
fn class_stmt() {
  check_module! {
    r#"#!hebi
      class T: pass
      class T:
        pass
      class T:
        fn f(v): pass
      class T:
        a = b
        fn f(v): pass
      class T(U): pass
      class T(U):
        pass
      class T(U):
        a = b
      class T(U):
        a = b
        fn f(v): pass
      class T(U):
        a = b
        fn f(v):
          pass
    "#
  }

  check_error! {
    r#"#!hebi
      class
        T: pass
    "#
  }

  check_error! {
    r#"#!hebi
      class T
        : pass
    "#
  }

  check_error! {
    r#"#!hebi
      class T: a = b
    "#
  }

  check_error! {
    r#"#!hebi
      class T: fn f(v): pass
    "#
  }

  check_error! {
    r#"#!hebi
      class T
        (U): pass
    "#
  }

  check_error! {
    r#"#!hebi
      class T(U)
        : pass
    "#
  }

  check_error! {
    r#"#!hebi
      class T(U):
        fn f(v): pass
        a = b
    "#
  }
}

#[test]
fn class_self_and_super() {
  check_module! {
    r#"#!hebi
      class T:
        fn f(self):
          print self
      
      class T(U):
        fn f(self):
          print self, super

      class T(U):
        init(self):
          self.v = super.f()
    "#
  }

  check_error! {
    r#"#!hebi
      self
    "#
  }

  check_error! {
    r#"#!hebi
      class T:
        v = self.f()
    "#
  }

  check_error! {
    r#"#!hebi
      fn f():
        print self
    "#
  }

  check_error! {
    r#"#!hebi
      class T:
        fn f():
          print self
    "#
  }

  check_error! {
    r#"#!hebi
      super
    "#
  }

  check_error! {
    r#"#!hebi
      fn f():
        print super
    "#
  }

  check_error! {
    r#"#!hebi
      class T:
        fn f():
          print self
    "#
  }

  check_error! {
    r#"#!hebi
      class T:
        v = super
    "#
  }

  check_error! {
    r#"#!hebi
      class T:
        fn f():
          print super
    "#
  }

  check_error! {
    r#"#!hebi
      class T(U):
        v = super
    "#
  }

  check_error! {
    r#"#!hebi
      class T(U):
        fn f():
          print super
    "#
  }
}

#[test]
fn duplicate_fields() {
  check_error! {
    r#"#!hebi
      class Test:
        a = 0
        a = 1
    "#
  }
  check_error! {
    r#"#!hebi
      class Test:
        a = 0
        fn a(): pass
    "#
  }
  check_error! {
    r#"#!hebi
      class Test:
        fn a(): pass
        fn a(): pass
    "#
  }
}

#[test]
fn whole_module() {
  check_module! {
    r#"#!hebi
      # variable declaration
      v := 0

      # values
      v = none # none
      v = 0.1 # number
      v = true # bool
      v = "\tas\\df\x2800\n" # string
      v = [none, 0.1, true, "\tas\\df\x2800\n"] # list
      v = {a: none, b: 0.1, c: true, d: "\tas\\df\x2800\n"} # table
      v = {["a"]: none, ["b"]: 0.1, ["c"]: true, ["d"]: "\tas\\df\x2800\n"}
      v = {[0]: none, [1]: 0.1, [2]: true, [3]: "\tas\\df\x2800\n"}

      # operators
      v = 2 + 2
      v = 2 - 2
      v = 2 / 2
      v = 2 * 2
      v = 2 % 2
      v = 2 ** 2
      v = 2 == 2
      v = 2 != 2
      v = 2 > 2
      v = 2 >= 2
      v = 2 < 2
      v = 2 <= 2
      v = -2
      v = !true
      v = true && true
      v = false || true
      v = a ?? b

      # assignment
      v = 1
      v += 1
      v -= 1
      v /= 1
      v *= 1
      v %= 1
      v **= 1
      v ??= 1

      # postfix
      v.a
      v["a"]
      v(a)

      # functions
      fn add(a, b):
        return a + b

      v = add(0, 1)

      fn fact(n):
        if n < 2:
          return n
        else:
          return n * fact(n - 1)

      fn print_fact(n):
        print(fact(n))

      # loops
      # range is an object
      for i in 0..10:
        print(i)

      # `yield` inside `fn` makes it an iterator
      # when called, iterators return an object with a `next` method
      # an iterator is done when its `next` method returns none
      fn counter(start, step, end):
        n := start
        loop:
          yield n
          n += step
          if end && n > end:
            return

      for n in counter(0, 10, 100):
        print(n)

      v = 0
      while v < 10:
        print(v)
        v += 1

      v = 0
      loop:
        if v >= 10:
          break
        print(v)
        v += 1

      if v < 10:
        print("less than 10")
      elif v < 20:
        print("less than 20")
      else:
        print("very large")

      class Test:
        init(self, n):
          self.n = n

        fn get_n(self):
          return self.n

        fn test1(self):
          print("instance", self)

        fn test0():
          print("static", Test)

      v = Test()
      print(v.get_n() == Test.get_n(v)) # true

      v = Test(10)

      Test.test0()
      v.test1()

      # errors
      # no exceptions, panic = abort
      panic("asdf")

      # modules
      import json

      v = json.parse("{\"a\":0, \"b\":1}")
      print(v) # { a: 0, b: 1 }

      # data class, implicit initializer
      class A:
        a = 100
        # init(self, a = 100):
        #   self.a = a

      print(A().a)   # 100
      print(A(10).a) # 10

      class B:
        a = 100
        init(self): # override the implicit initializer
          pass

      print(B().a) # 100
      print(B(10).a) # error

      class C:
        # fields do not have to be declared
        # and may be added in the initializer
        # after `init` is called, the class is frozen
        # no fields/methods may be added or removed
        init(self):
          self.a = 10

      print(C().a) # 10
      C().b = 10 # error: cannot add new field `b` to class `C`

      class A:
        fn inherited(self):
          print("test 0")

      class B(A): pass

      A().inherited() # test 0
      B().inherited() # test 0

      class C(B):
        fn inherited(self): # override
          print("test 1")

      C().inherited() # test 1

      class D(C):
        fn inherited(self): # override with call to super
          super.inherited()
          print("test 2")

      D().inherited() # test 1
                      # test 2

      class X:
       init(self):
          self.v = 10

      class Y(X):
        init(self): # error: `super.init` must be called before accessing `self` or returning in derived constructor
          self.v = 10

      class Z(X):
        init(self, v):
          super.init()
          self.v += v

      print(Z(15).v) # 25
    "#
  }
}

/* #[test]
fn _temp() {
  check_error! {
    r#"#!hebi
      class T: fn f(v): pass
    "#
  }
} */

// semicolons

check_module! {
  many_semicolons,
  r#"#!hebi
    1 + 2; 3 + 4
    1 + 2; 3 + 4;
    print 5; print 6;
    pass; print 7; pass;
    a; b; c; d;
    a + b - c; a / d || c; print abcd;
    print (
      a + b
    ); print (
        c + d
      ); print (
          e + f
        );  
    if true: print x; print y;; print z; if false: print a;; print b
    if true: print x; print y;; elif false: print a;; else: print z; print zz
    if one: print one; if two: print two;; else: print two_else;;;; else: print one_else
  "#
}

check_module! {
  if_statement_supports_semicolons,
  r#"#!hebi
    if true:
      if true: print x; print y
      if true: print z
  "#
}

check_module! {
  for_statement_supports_semicolons,
  r#"#!hebi
    for i in 0..10: print x
    x := 0; for i in 0..10: x += i; print x;; print x;
    "#
}

check_module! {
  for_statements_can_be_nested_on_the_same_line,
  r#"#!hebi
    for y in 0..10: for x in 0..10: print "x * y =", x * y;; print "------------"
  "#
}

check_module! {
  infinite_loop_statements_support_semicolons,
  r#"#!hebi
    loop: print x; print y
    loop: print x;; print y
    loop: print x; print y;; print z
  "#
}

check_module! {
  while_statements_support_semicolons,
  r#"#!hebi
    x := 1; while x % 96 != 0: x += x * 17; x += x * 11;; if x != 46656: print "broken arithmetics"
  "#
}

check_module! {
  import_statements_support_semicolons,
  r#"#!hebi
    import a; import b
    from a import b; import x; from c import d;
    import http; from json import parse; print parse(http.get("https://jsonplaceholder.typicode.com/todos/1"))
  "#
}

check_module! {
  trailing_statements_of_nested_blocks_are_included_in_outer_block_body,
  r#"#!hebi
    while outer: print "outer"; while inner: break;; print "after inner";
  "#
}

check_module! {
  fn_statement_supports_semicolons,
  r#"#!hebi
    fn add(a, b): s := a + b; print "a + b = ", s; return s;; print add(1,2); 
    fn reduce(it, init, func): acc := init; for x in it: acc = func(acc, x);; return acc;; fn reducer(acc, i): return acc + i;; print reduce([1,2,3,4],0, reducer)
  "#
}

check_error! {
  inline_for_indentation_enforcement,
  r#"#!hebi
    for i in 0..10: print x;
      print y

    for y in 0..10: for x in 0..10: print "x * y =", x * y;
        print x, y

    for y in 0..10: for x in 0..10: print "x * y =", x * y;;
      print "---------"
  "#
}

check_error! {
  inline_while_indentation_enforcement,
  r#"#!hebi
    while x % 96 != 0: x += x * 17; x += x * 11;
      if x != 46656: print "broken arithmetics"
  "#
}

check_module! {
  funky_inline_indentation_mix,
  r#"#!hebi
    if cond: print 1;; else:
      print "i == j"

    for i in (call()
    ): for j in (call()
      ): if (
          i 
          != 
          j
        ): print (
            i, 
            j
          ); print (
            j, 
            i
          );; else: 
          print (
          "i == j"
        )

    for one in x(): for two in (
      y()
    ): for three in z(): if (one != two
    ): print one, two;; elif one == two: print (
      "one == two"
                          );; else: 
    print "unreachable"
  "#
}

check_error! {
  scope_no_semi,
  r#"#!hebi
    if true: print x
      print y
  "#
}

check_error! {
  scope_single_semi,
  r#"#!hebi
    if true: print x;
      print y
  "#
}

check_error! {
  scope_double_semi,
  r#"#!hebi
    if true: print x;;
      print y
  "#
}

check_error! {
  scope_nested_no_semi,
  r#"#!hebi
    if true: if true: print x
      print y
  "#
}

check_error! {
  scope_nested_single_semi,
  r#"#!hebi
    if true: if true: print x;
      print y
  "#
}

check_error! {
  scoped_nested_double_semi,
  r#"#!hebi
    if true: if true: print x;;
      print y
  "#
}
