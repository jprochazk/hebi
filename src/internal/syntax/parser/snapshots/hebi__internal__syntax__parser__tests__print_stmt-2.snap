---
source: src/internal/syntax/parser/tests.rs
expression: errors
---
invalid indentation
| [4;31m"a"[0m


